### PR TITLE
ENH: Support multiple metrics

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/dask/dask-searchcv/issues/52

```python
In [1]: from sklearn.datasets import make_hastie_10_2
   ...: from sklearn.metrics import make_scorer
   ...: from sklearn.metrics import accuracy_score
   ...: from sklearn.tree import DecisionTreeClassifier
   ...:
   ...: import dask_searchcv as dcv
   ...:
   ...: X, y = make_hastie_10_2(n_samples=1000, random_state=42)
   ...:
   ...: # The scorers can be either be one of the predefined metric strings or a scorer
   ...: # callable, like the one returned by make_scorer
   ...: scoring = {'AUC': 'roc_auc', 'Accuracy': make_scorer(accuracy_score)}
   ...:
   ...: # Setting refit='AUC', refits an estimator on the whole dataset with the
   ...: # parameter setting that has the best cross-validated AUC score.
   ...: # That estimator is made available at ``gs.best_estimator_`` along with
   ...: # parameters like ``gs.best_score_``, ``gs.best_parameters_`` and
   ...: # ``gs.best_index_``
   ...: a = dcv.GridSearchCV(DecisionTreeClassifier(random_state=42),
   ...:                      param_grid={'min_samples_split': range(2, 403, 10)},
   ...:                      scoring=scoring, cv=5, refit='AUC')
   ...: a.fit(X, y)
   ...:

In [2]: a.scorer_
Out[2]:
{'AUC': make_scorer(roc_auc_score, needs_threshold=True),
 'Accuracy': make_scorer(accuracy_score)}

In [3]: a.multimetric_
Out[3]: True

In [4]: a.cv_results_.keys()
Out[4]: dict_keys(['params', 'mean_fit_time', 'std_fit_time', 'mean_score_time',
... 'std_score_time', 'split0_test_AUC', 'split1_test_AUC', 'split2_test_AUC',
... 'split3_test_AUC', 'split4_test_AUC', 'mean_test_AUC', 'std_test_AUC',
... 'rank_test_AUC', 'split0_test_Accuracy', 'split1_test_Accuracy',
... 'split2_test_Accuracy', 'split3_test_Accuracy', 'split4_test_Accuracy',
... 'mean_test_Accuracy', 'std_test_Accuracy', 'rank_test_Accuracy',
... 'split0_train_AUC', 'split1_train_AUC', 'split2_train_AUC',
... 'split3_train_AUC', 'split4_train_AUC', 'mean_train_AUC',
... 'std_train_AUC', 'split0_train_Accuracy', 'split1_train_Accuracy',
... 'split2_train_Accuracy', 'split3_train_Accuracy', 'split4_train_Accuracy',
... 'mean_train_Accuracy', 'std_train_Accuracy', 'param_min_samples_split'])
```

I'll add more tests later, but does the overall approach look OK here @jcrist? I had to modify a few places, but none of the modifications were that extensive.

cc @jnothman

### CODE REVIEW

**Comment by jnothman** [comment]
We don't exactly make `*SearchCV` compatibility easy, do we? But I guess we don't want the burden of maintaining public versions of `_fit_and_score` etc.

File: dask_searchcv/methods.py [hunk 1]
```diff
@@ -321,13 +330,23 @@ def create_cv_results(scores, candidate_params, n_splits, error_score, weights):
...
-    if train_scores is not None:
-        _store(results, 'train_score', train_scores,
-               n_splits, n_candidates, splits=True)
+
+    if not multimetric:
+        _store(results, 'test_score', test_scores, n_splits, n_candidates,
+            splits=True, rank=True, weights=weights)
+        if train_scores is not None:
+            _store(results, 'train_score', train_scores,
+                n_splits, n_candidates, splits=True)
```
Comment (jcrist) [inline dask_searchcv/methods.py]: Nit: indentation.

File: dask_searchcv/methods.py [hunk 2]
```diff
@@ -343,8 +362,11 @@ def create_cv_results(scores, candidate_params, n_splits, error_score, weights):
     return results
 
 
-def get_best_params(candidate_params, cv_results):
-    best_index = np.flatnonzero(cv_results["rank_test_score"] == 1)[0]
+def get_best_params(candidate_params, cv_results, scorer):
+    if scorer is True:
+        scorer = 'score'
```
Comment (jcrist) [inline dask_searchcv/methods.py]: This took me a bit to figure out, would prefer if this logic was moved out of the function, and `'score'` was passed in the call instead.

File: dask_searchcv/model_selection.py [hunk 3]
```diff
@@ -678,6 +681,10 @@ def __init__(self, estimator, scoring=None, iid=True, refit=True, cv=None,
         self.scheduler = scheduler
         self.n_jobs = n_jobs
         self.cache_cv = cache_cv
+        if self.refit in (True, False):
+            self._score_key = 'score'
+        else:
+            self._score_key = self.refit
```
Comment (jcrist) [inline dask_searchcv/model_selection.py]: Would prefer not to store hidden state on the class, as that's an antipattern in scikit-learn. Could implement this as a property if it cleans up the code.

File: dask_searchcv/model_selection.py [hunk 4]
```diff
@@ -769,7 +776,26 @@ def fit(self, X, y=None, groups=None, **fit_params):
...
+        # This is {None, True, False}
+        # None indicates scikit-learn < 0.19 (when multimetric was added)
+        # True indicates multiple metrics were used
+        # False indicates a single metric was used
+        multimetric = getattr(self, 'multimetric_', None)
+        if multimetric:
+            multimetric = self.scorer_.keys()
+        elif multimetric is False:
+            # sklearn >= 0.19.0, but single metric
+            self.scorer_ = self.scorer_['score']
```
Comment (jcrist) [inline dask_searchcv/model_selection.py]: I think this is slightly cleaner/clearer:

```python
if _HAS_MULTIPLE_METRICS:
    from sklearn.metrics.scorer import _check_multimetric_scoring
    
    scorer, multimetric = _check_multimetric_scoring(estimator, scoring=self.scoring)
    if not multimetric:
        scorer = scorer['score']
    # Attribute only on sklearn >= 0.19.0
    self.multimetric_ = multimetric
else:
    scorer = check_scoring(estimator, scoring=scoring)
    multimetric = False

self.scorer_ = scorer
```

Then pass the boolean `multimetric` to `build_graph`, where the list of score names will be made if needed. I think this keeps the necessary logic localized, and also avoids reassigning variables with a conceptually different value.

File: dask_searchcv/tests/test_model_selection.py [hunk 5]
```diff
@@ -620,3 +621,39 @@ def test_scheduler_param_distributed(loop):
 def test_scheduler_param_bad():
     with pytest.raises(ValueError):
         _normalize_scheduler('threeding', 4)
+
+
+@pytest.mark.skipif(not _HAS_MULTIPLE_METRICS, reason="Added in 0.19.0")
+def test_multiple_metrics():
```
Comment (jcrist) [inline dask_searchcv/tests/test_model_selection.py]: Could also use a test with fit failures. There are a few examples in this file and `test_model_selection_sklearn.py`.

Also, if this test is borrowed from sklearn (or if there are tests using this feature that we could borrow from sklearn) then they should go in `test_model_selection_sklearn.py`.

File: dask_searchcv/model_selection.py [hunk 6]
```diff
@@ -106,12 +108,13 @@ def build_graph(estimator, cv, scorer, candidate_params, X, y=None,
     candidate_params_name = 'cv-parameters-' + main_token
     dsk[candidate_params_name] = (decompress_params, fields, params)
     dsk[cv_results] = (create_cv_results, scores, candidate_params_name,
-                       n_splits, error_score, weights)
+                       n_splits, error_score, weights, multimetric)
```
Comment (jcrist) [inline dask_searchcv/model_selection.py]: Following my comment on `multimetric` in the `fit` call, this would then `metrics` instead of `multimetric`, as follows:

```
if multimetric:
    metrics = list(scorer.keys())
else:
    metrics = None
```

**Comment by TomAugspurger** [comment]
Thanks @jcrist much cleaner now I think.

I'm having trouble propagating failures with multiple metrics, so the CI will fail for now. I'll push a fix tomorrow morning.

**Comment by TomAugspurger** [comment]
OK, I think I've got it now.

One question I had is how to handle `best_score_`, `best_index_`, `best_params_`, etc. when `refit=False`. I see two options:

1. Don't set them (this would be consistent with scikit-learn, but not dask-searchcv)
2. Make them dictionaries of `{scorer -> float / int / dict}` with the best score / index / params for that particular scorer.

I'm going to try out 2, and see what it breaks.

**Comment by TomAugspurger** [comment]
FYI all green @jcrist.

I'm no 100% convinced that the way I've handled the `best_*` is the best way of doing things. It seems weird for the return type of a property to depend on *two* parameters (scoring being a dict, and refit being `False`). Still, I haven't come up with a cleaner way.

**Comment by TomAugspurger** [comment]
Ok, hopefully things are good now: Changes:

The `test_no_refit` test was incorrectly passing since the `assert` was on *any* of the conditions being true, instead of *all*. I fixed that test (and the implemention) in https://github.com/dask/dask-searchcv/pull/60/commits/9c677329477bdf692c584acc5d51281fd17cfcf7 The `best_*` properties will still show up in tab completion. I'm inclined to ignore this though. (Side-note, I didn't know that if you raised `AttributeError` in a property, then `hasattr(thing, attr)` is False, which is neat).

That cleaned up the multimetric code quite a bit, since my optionally returning a `dict` if `refit` is False moot.

Finally, I've borrowed scikit-learn's behavior / error message when fitting with multiple metrics and `refit=True`.

**Comment by TomAugspurger** [comment]
And here's a gist with the attr types of sklearn / dask-searchcv for various combos of scoring / refit.

**Comment by PeterDSteinberg** [comment]
@TomAugspurger  - Glad you're working on support for multiple metrics.  In [Elm](https://github.com/ContinuumIO/elm) we need this to support multiobjective optimization in `elm.model_selection.EaSearchCV`.

**Comment by TomAugspurger** [comment]
👍 , I think this is all ready @jcrist, if you have a chance to glance through the implementation again.

**Comment by jcrist** [comment]
Apologies for the delay in response:

> I'm no 100% convinced that the way I've handled the best_* attributes is the best way of doing things. It seems weird for the return type of a property to depend on two parameters (scoring being a dict, and refit being False). Still, I haven't come up with a cleaner way.

I'd prefer to 100% match newer sklearn (0.19.1) behavior with respect to set attributes and handling of scorer/refit. I think this should be possible without to much effort. In this case we'd only set `best_score_` with multiple metrics, etc... if `refit` is set:

```
    refit : boolean, or string, default=True
        Refit an estimator using the best found parameters on the whole
        dataset.

        For multiple metric evaluation, this needs to be a string denoting the
        scorer is used to find the best parameters for refitting the estimator
        at the end.

        The refitted estimator is made available at the ``best_estimator_``
        attribute and permits using ``predict`` directly on this
        ``GridSearchCV`` instance.

        Also for multiple metric evaluation, the attributes ``best_index_``,
        ``best_score_`` and ``best_parameters_`` will only be available if
        ``refit`` is set and all of them will be determined w.r.t this specific
        scorer.

        See ``scoring`` parameter to know more about multiple metric
        evaluation.
```

In general I'd like `dask-searchcv` to match the behavior of whatever `scikit-learn` version is installed as much as possible.

**Comment by TomAugspurger** [comment]
> I'd prefer to 100% match newer sklearn (0.19.1) behavior with respect to set attributes and handling of scorer/refit. 

I believe this branch currently achieves that. The only difference is that since `best_params_` and `best_score_` are properties, they show up in `dir(cls)`, but I think that's OK.

```python
In [20]: import sklearn.model_selection as ms; import dask_searchcv as dcv

In [21]: from sklearn.ensemble import RandomForestClassifier; from sklearn.datasets import make_classification

In [22]: X, y = make_classification()

In [23]: a = dcv.GridSearchCV(RandomForestClassifier(), param_grid={'n_estimators': [10, 100]}, scoring={"auc": "roc_auc", "Accuracy": 'accuracy'}, refit=Fals
    ...: e)

In [24]: b = ms.GridSearchCV(RandomForestClassifier(), param_grid={'n_estimators': [10, 100]}, scoring={"auc": "roc_auc", "Accuracy": 'accuracy'}, refit=False
    ...: )

In [25]: a.fit(X, y); b.fit(X, y)
Out[25]:
GridSearchCV(cv=None, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_estimators': [10, 100]}, pre_dispatch='2*n_jobs',
       refit=False, return_train_score='warn',
       scoring={'auc': 'roc_auc', 'Accuracy': 'accuracy'}, verbose=0)

In [26]: a.best_params_
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-26-ab762e88e145> in <module>()
----> 1 a.best_params_

~/Envs/dask-dev/lib/python3.6/site-packages/dask-searchcv/dask_searchcv/model_selection.py in best_params_(self)
    704     def best_params_(self):
    705         check_is_fitted(self, 'cv_results_')
--> 706         self._check_if_refit('best_params_')
    707         return self.cv_results_['params'][self.best_index_]
    708

~/Envs/dask-dev/lib/python3.6/site-packages/dask-searchcv/dask_searchcv/model_selection.py in _check_if_refit(self, attr)
    695         if not self.refit:
    696             raise AttributeError(
--> 697                 "'{}' is not a valid attribute with 'refit=False'.".format(attr))
    698
    699     @property

AttributeError: 'best_params_' is not a valid attribute with 'refit=False'.

In [27]: b.best_params_
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-27-9bb303929fe5> in <module>()
----> 1 b.best_params_

AttributeError: 'GridSearchCV' object has no attribute 'best_params_'
```

The error message is different, though ours is a bit better :) This is tested in https://github.com/dask/dask-searchcv/pull/60/files#diff-6980992e8dfcc62386acb240d0d401cdR986

**Comment by TomAugspurger** [comment]
Oh, and to be clear, that "I'm not 100% sure" comment from https://github.com/dask/dask-searchcv/pull/60#issuecomment-335919394 was based on an old implementation. I'm much more confident that we match scikit-learn's behavior now.

File: dask_searchcv/model_selection.py [hunk 7]
```diff
@@ -843,11 +898,20 @@ def visualize(self, filename='mydask', format=None, **kwargs):
...
+    (see :ref:`scoring`) to evaluate the predictions on the test set.
+
+    For evaluating multiple metrics, either give a list of (unique) strings
+    or a dict with names as keys and callables as values.
+
+    NOTE that when using custom scorers, each scorer should return a single
+    value. Metric functions returning a list/array of values can be wrapped
+    into multiple scorers that return one value each.
+
+    See :ref:`multimetric_grid_search` for an example.
```
Comment (jcrist) [inline dask_searchcv/model_selection.py]: These refs don't exist in our docs, please remove them.

File: dask_searchcv/methods.py [hunk 8]
```diff
@@ -321,13 +335,23 @@ def create_cv_results(scores, candidate_params, n_splits, error_score, weights):
...
+               splits=True, rank=True, weights=weights)
+        if train_scores is not None:
+            _store(results, 'train_score', train_scores,
+                   n_splits, n_candidates, splits=True)
+    else:
+        for key in multimetric:
+            _store(results, 'test_{}'.format(key), test_scores[key], n_splits,
+                   n_candidates, splits=True, rank=True, weights=weights)
+        for key in multimetric:
+            if train_scores is not None:
```
Comment (jcrist) [inline dask_searchcv/methods.py]: Should swap check and loop.

File: dask_searchcv/model_selection.py [hunk 9]
```diff
@@ -769,7 +792,33 @@ def fit(self, X, y=None, groups=None, **fit_params):
...
+                        # This will work for both dict / list (tuple)
+                        self.refit not in scorer):
+                    raise ValueError("For multi-metric scoring, the parameter "
+                                    "refit must be set to a scorer key "
+                                    "to refit an estimator with the best "
+                                    "parameter setting on the whole data and "
+                                    "make the best_* attributes "
+                                    "available for that metric. If this is not "
+                                    "needed, refit should be set to False "
+                                    "explicitly. %r was passed." % self.refit)
```
Comment (jcrist) [inline dask_searchcv/model_selection.py]: Nit: indentation.

File: dask_searchcv/methods.py [hunk 10]
```diff
@@ -321,13 +335,23 @@ def create_cv_results(scores, candidate_params, n_splits, error_score, weights):
...
+                _store(results, 'train_{}'.format(key), train_scores[key], n_splits,
+                    n_candidates, splits=True)
```
Comment (jcrist) [inline dask_searchcv/methods.py]: Nit: indentation.

File: dask_searchcv/model_selection.py (hunk 7 above)
Comment (TomAugspurger) [reply dask_searchcv/model_selection.py]: Happy to remove them, but do you want me to setup intersphinx so that they do exist?

Comment (jcrist) [reply dask_searchcv/model_selection.py]: Up to you, I've never used intersphinx before. If you think it aids in clarity, then sure - perhaps tweak the surrounding language to indicate this is a scikit-learn example/reference instead of a `dask-searchcv` example?

Comment (TomAugspurger) [reply dask_searchcv/model_selection.py]: Just removed for now.

**Comment by jcrist** [comment]
Looks good, merging. Thanks!
